core/authenticate/service.go

```diff
@@ -316,6 +385,208 @@ func (s Service) applyMailOTP(ctx context.Context, request RegistrationFinishReq
+		return nil, err
+	}
+
+	session.Challenge = base64.RawURLEncoding.EncodeToString([]byte(session.Challenge))
+	sessionInBytes, err := json.Marshal(session)
+	if err != nil {
+		return nil, err
+	}
+	flow.Metadata["passkey_session"] = sessionInBytes
+	flow.Metadata["passkey_type"] = "register"
+	if err = s.flowRepo.Set(ctx, flow); err != nil {
+		return nil, err
+	}
+	optionsBytes, err := json.Marshal(options)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RegistrationStartResponse{
+		Flow:  flow,
+		State: flow.ID.String(),
+		StateConfig: map[string]any{
+			"options": optionsBytes,
+		},
+	}, nil
+}
+
+func (s Service) startPassKeyLoginMethod(ctx context.Context, loggedInUser user.User, flow *Flow) (*RegistrationStartResponse, error) {
+	newPassKeyUser := strategy.NewPassKeyUser(flow.Email)
+	decodedCredBytes, err := base64.StdEncoding.DecodeString(loggedInUser.Metadata["passkey_credentials"].(string))
+	if err != nil {
+		return nil, err
+	}
+
+	var webAuthCredentialData []webauthn.Credential
+	err = json.Unmarshal(decodedCredBytes, &webAuthCredentialData)
+	if err != nil {
+		return nil, err
+	}
+	newPassKeyUser.Credentials = webAuthCredentialData
+	options, session, err := s.webAuth.BeginLogin(newPassKeyUser)
+	if err != nil {
+		return nil, err
+	}
+
+	session.Challenge = base64.RawURLEncoding.EncodeToString([]byte(session.Challenge))
+	sessionInBytes, err := json.Marshal(session)
+	if err != nil {
+		return nil, err
+	}
+	flow.Metadata["passkey_session"] = sessionInBytes
+	flow.Metadata["passkey_type"] = "login"
+	if err = s.flowRepo.Set(ctx, flow); err != nil {
+		return nil, err
+	}
+	optionsBytes, err := json.Marshal(options)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RegistrationStartResponse{
+		Flow:  flow,
+		State: flow.ID.String(),
+		StateConfig: map[string]any{
+			"options": optionsBytes,
+		},
+	}, nil
+}
+
+func (s Service) finishPassKeyRegisterMethod(ctx context.Context, request RegistrationFinishRequest) (*RegistrationFinishResponse, error) {
+	passkeyOptions, ok := request.StateConfig["options"].(string)
+	if !ok {
+		return nil, errors.New("invalid auth state")
+	}
+	requestReader := bytes.NewReader([]byte(passkeyOptions))
+	credentialCreationResponse, err := protocol.ParseCredentialCreationResponseBody(requestReader)
+	if err != nil {
+		return nil, err
+	}
+
+	flowIdString := request.State
+	flowId, err := uuid.Parse(flowIdString)
+	if err != nil {
+		return nil, err
+	}
+	flow, err := s.flowRepo.Get(ctx, flowId)
+	if err != nil {
+		return nil, err
+	}
+
+	userFinishRegister := strategy.NewPassKeyUser(flow.Email)
+	sessionInterface := flow.Metadata["passkey_session"]
+	var webAuthSessionData webauthn.SessionData
+	sessionBytes, err := base64.StdEncoding.DecodeString(sessionInterface.(string))
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(sessionBytes, &webAuthSessionData)
+	if err != nil {
+		return nil, err
+	}
+	userFinishRegister.Id = string(webAuthSessionData.UserID)
```

Comment:
Why not create the user(userFinishRegister := strategy.NewPassKeyUser(flow.Email)) here only and pass `UserID` as constructor?